# 
# This patch contains bug fixes to the PAM credential and session operations.
# In the original OpenSSH, the server only gives warnings and still allows 
# users to log in when pam_setcred() or pam_open_session() fail, if user auth
# method is not keyboard-interactive or password.  This is not a correct
# behavior. The server should just fatal out, when these functions fail.
#
# We have contributed back these bug fixes to the OpenSSH upstream community. 
# For more information, see https://bugzilla.mindrot.org/show_bug.cgi?id=2399
# In the future, if these bug fixes are accepted by the upstream in a later
# release, we will remove this patch when we upgrade to that release.
#
--- openssh-9.4p1/auth-pam.c
+++ openssh-9.4p1/auth-pam.c
@@ -1246,12 +1246,19 @@ if (sshpam_err == PAM_SUCCESS) {
 		sshpam_cred_established = 1;
 		return;
 	}
+
+#ifdef PAM_BUGFIX
+	/* Server will fatal out when pam_setcred() failed. */
+	fatal("PAM: pam_setcred(): %s", pam_strerror(sshpam_handle,
+	    sshpam_err));
+#else /* orig */
 	if (sshpam_authenticated)
 		fatal("PAM: pam_setcred(): %s",
 		    pam_strerror(sshpam_handle, sshpam_err));
 	else
 		debug("PAM: pam_setcred(): %s",
 		    pam_strerror(sshpam_handle, sshpam_err));
+#endif /* PAM_BUGFIX */
 }
 
 static int
@@ -1347,10 +1354,16 @@ sshpam_err = pam_open_session(sshpam_han
 	if (sshpam_err == PAM_SUCCESS)
 		sshpam_session_open = 1;
 	else {
+#ifdef PAM_BUGFIX
+		/* Server will fatal out when pam_open_session() failed */
+		fatal("PAM: pam_open_session(): %s",
+		    pam_strerror(sshpam_handle, sshpam_err));
+#else /* orig */
 		sshpam_session_open = 0;
 		auth_restrict_session(ssh);
 		error("PAM: pam_open_session(): %s",
 		    pam_strerror(sshpam_handle, sshpam_err));
+#endif /* PAM_BUGFIX */
 	}
 
 }
